enhancements/project-request-limit.md

```diff
@@ -0,0 +1,392 @@
+namespace. At a high level, the call chain looks as below:
+```
+    Project request -> openshift-apiserver -> kube-apiserver -> ProjectRequestLimit webhook.
+```
+
+If we do a deep dive, the call chain to the admission webhook looks as below:
+```
+Project request -> kube-apiserver -> kube aggregator layer inside the kube-apiserver ->
+   openshift-apiserver -> admission layer in openshift-apiserver -> kube-apiserver -> 
+     kube-aggregator layer inside the kube-apiserver -> ProjectRequestLimit webhook 
+```
+
+The webhook server will have the following topology:
+* The webhook is deployed as a `DaemonSet` server.
+* We restrict the webhook server to the master nodes only. Apply appropriate `nodeSelector` and `tolerations` to the 
+  `PodSpec` so that the `Pods` are scheduled on to the master nodes.
+
+
+#### The Operator
+The operator will allow a cluster admin to:
+* Enable or disable `ProjectRequestLimit` admission webhook on a cluster.
+* Specify `limits` on `Project` create request(s) that the `ProjectRequestLimit` admission webhook can enforce.
+* Manage other lifecycle aspects of the operand.
+
+The operator will define a CRD to expose its API. The cluster admin will interact with the operator via a custom resource:
+* We treat the `ProjectRequestLimit` admission webhook (represented by a `DaemonSet`) as a cluster singleton. That mean
+  the operator needs to manage a single instance of the `DaemonSet`.
+* For the above reason, the CRD will be defined as `cluster-scoped`.
+* The operator will be reconciling a CR named `cluster`, it will ignore other custom resources.
+
+```
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: projectrequestlimits.operator.project.openshift.io
+spec:
+  group: operator.project.openshift.io
+  version: v1
+  scope: Cluster
+  names:
+    plural: projectrequestlimits
+    singular: projectrequestlimit
+    kind: ProjectRequestLimit
+```  
+
+This is how a cluster-admin can enable the `ProjectRequestLimit` admission webhook on a cluster with `limits` to be enforced. 
+```
+apiVersion: operator.project.openshift.io/v1
+kind: ProjectRequestLimit
+metadata:
+  name: cluster
+spec:
+  limits:
+  // for selector level=admin, no maxProjects is specified. This means that users with this label 
+  // will not have a maximum of project requests.
+  - selector:
+      level: admin
+  
+  //  for selector level=advanced, a maximum number of 10 projects will be allowed.
+  - selector:
+      level: advanced
+    maxProjects: 10
+  
+  // no selector is specified. This means that it will be applied to any user that doesn’t satisfy 
+  // the previous two rules. Because rules are evaluated in order, this rule should be specified last.
+  - maxProjects: 2
+```
+
+##### Install:
+* The operator is installed into a predefined namespace by OLM: The predefined namespace `openshift-project-request-limit-operator` 
+  will be wired in to the OLM manifests.
```

Comment:
wondering whether we want to make openshift-* namespaces available to OLM component. @deads2k had opinions

Reply:
All OpenShift clusters have a predefined `openshift-operators` namespace where a cluster admin can install OLM enabled operators. I am not sure if we want to mingle `ProjectReuestLimit`  with all other OLM enabled operators.

should each operator we provide be in its own namespace? It sounds better from an isolation perspective. or Is it worth creating a dedicated namespace (via cvo) for non critical webhooks `openshift-admission-operators` (we can find a better names )?